src/deploy/functions/params.ts

```diff
@@ -0,0 +1,342 @@
+  input?: TextInput<string> | SelectInput<string>;
+}
+
+export interface IntParam extends ParamBase<number> {
+  type: "int";
+
+  // If omitted, defaults to TextInput<number>
+  input?: TextInput<number> | SelectInput<number>;
+}
+
+export interface TextInput<T, Extensions = {}> { // eslint-disable-line
+  type?: "text";
+
+  text:
+    | Extensions
+    | {
+        example?: string;
+      };
+}
+
+interface SelectOptions<T> {
+  // Optional human-facing value for this option (e.g. "US Central (Iowa)" instead of value
+  // "us-central1")
+  label?: string;
+
+  // Actual value of the parameter if this option is selected
+  value: T;
+}
+
+export interface SelectInput<T> {
+  type?: "select";
+
+  select: Array<SelectOptions<T>>;
+}
+
+export type Param = StringParam | IntParam;
+type ParamValue = string | number | boolean;
+
+/**
+ * Calls the corresponding resolveX function for the type of a param.
+ * To be used when resolving the default value of a param, if CEL.
+ * It's an error to call this on a CEL expression that depends on params not already known in the currentEnv.
+ */
+function resolveDefaultCEL(
+  type: string,
+  expr: CEL,
+  currentEnv: Record<string, ParamValue>
+): ParamValue {
+  const deps = dependenciesCEL(expr);
+  const allDepsFound = deps.every((dep) => {
+    return Object.keys(currentEnv).includes(dep);
+  });
+  if (!allDepsFound) {
+    throw new FirebaseError(
+      "Build specified parameter with un-resolvable default value " +
+        expr +
+        "; dependencies missing."
+    );
+  }
+
+  switch (type) {
+    case "string":
+      return resolveString(expr, currentEnv);
+    case "int":
+      return resolveInt(expr, currentEnv);
+    default:
+      throw new FirebaseError(
+        "Build specified parameter with default " + expr + " of unsupported type"
+      );
+  }
+}
+/**
+ * Tests whether a mooted ParamValue literal is of the correct type to be the value for a Param.
+ */
+function canSatisfyParam(param: Param, value: ParamValue): boolean {
+  if (param.type === "string") {
+    return typeof value === "string";
+  } else if (param.type === "int") {
+    return typeof value === "number" && Number.isInteger(value);
+  } else {
+    assertExhaustive(param);
+  }
+}
+
+/**
+ * A param defined by the SDK may resolve to:
+ * - the value of a Cloud Secret in the same project with name == param name (not implemented yet), but only if it's a SecretParam
+ * - a literal value of the same type already defined in one of the .env files with key == param name
+ * - the value returned by interactively prompting the user
+ *   - the default value of the prompt comes from the SDK via param.default, which may be a literal value or a CEL expression
+ *   - if the default CEL expression is not resolvable--it depends on a param whose value is not yet known--we throw an error
+ *   - yes, this means that the same set of params may or may not throw depending on the order the SDK provides them to us in
```

Comment:
Or, more interestingly, the value of one param can set the defaults of many params. The sendgrid extension, for example, asks for a collection name three different times for three different features and then asks you for a key inside that collection. Most people probably have a single collection with three keys in it. Using this behavior the customer can say:

```
const coll1 = params.defineString("COLL_1", { description: "The collection where you do foo" });
const coll2 = params.defineString("COLL_2", { description: "The collection where you bar", default: coll1 });
const coll3 = params.defineString("COLL_3", { description: "The collection where you baz", default: coll2 });
```

Then the installer saves typing in a collection name up to twice.